proofread: s3_gateway.md (#2016)
juicedata/juicefs: docs/en/deployment/s3_gateway.md

---
# Deploy JuiceFS S3 Gateway

> JuiceFS has introduced S3 gateway since v0.11. It's a feature implemented based on the [MinIO S3 Gateway](https://docs.min.io/docs/minio-gateway-for-s3.html). It provides an S3-compatible RESTful API for files on JuiceFS, enabling the management of files stored on JuiceFS with tools such as s3cmd, AWS CLI, MinIO Client (mc) etc., in cases where mounting is not convenient. In addition, S3 gateway also provides a web-based file manager that allows users to manage files in web browsers.

Reviewer: ```suggestion
JuiceFS has introduced S3 gateway since v0.11. The feature is implemented based on the [MinIO S3 Gateway](https://docs.min.io/docs/minio-gateway-for-s3.html). It provides an S3-compatible RESTful API for files on JuiceFS, enabling the management of files stored on JuiceFS with tools such as s3cmd, AWS CLI, and MinIO Client (mc) in cases where mounting is not convenient. In addition, S3 gateway also provides a web-based file manager that allows users to manage files in web browsers.
```

> Since JuiceFS stores files in chunks, the files cannot be accessed directly through the interfaces of the underlying object storage. The S3 gateway provides access similar to the underlying object storage, with the following architecture diagram.

Reviewer: ```suggestion
Since JuiceFS stores files in chunks, the files cannot be accessed directly through the interfaces of the underlying object storage. The S3 gateway accesses the underlying object storage in a similar way, shown in the following architecture diagram.
```
不知道意思变了没？

> - Third-party clients in the host where the S3 gateway is located can use `http://127.0.0.1:9000` or `http://localhost:9000` for access.

Reviewer: ```suggestion
- A third-party client in the host where the S3 gateway is located can use `http://127.0.0.1:9000` or `http://localhost:9000` for access.
```

- A third-party client on the same LAN as the host where the S3 gateway is located can access it using `http://192.168.1.8:9000` (assuming the intranet IP address of the S3 gateway-enabled host is 192.168.1.8).
- The S3 gateway can be accessed over the Internet using `http://110.220.110.220:9000` (assuming that the public IP address of the S3 gateway-enabled host is 110.220.110.220).

> ## Accessing S3 gateway

Reviewer: 上一层标题 `Deploy JuiceFS S3 Gateway` 用的是动词，但是这个`subheading` 用的都是现代分词，是不是需要统一一下？

Default output format [None]:
```

> The program will guide you through adding the new configuration interactively, where `Access Key ID` is the same as `MINIO_ROOT_USER` and `Secret Access Key` is the same as `MINIO_ROOT_PASSWORD`. Leave the region name and output format blank.

Reviewer: ```suggestion
The program will guide you interactively to add the new configuration, where `Access Key ID` and `Secret Access Key` are the same as `MINIO_ROOT_USER` and `MINIO_ROOT_PASSWORD`, respectively. The region name and output format should be left blank.
```

$ mc alias set juicefs http://localhost:9000 admin 12345678 --api S3v4
```

> Following the mc command format, the above command creates a configuration with the alias `juicefs`. In particular, note that the API version `-api "s3v4"` must be specified in the command.

Reviewer: ```suggestion
Following the mc command format, the above command creates a configuration with the alias `juicefs`. Note that the API version `-api S3v4` must be specified in the command.
```

> Then, you can freely manage the copying, moving, adding and deleting of files and folders between your local disk and JuiceFS storage as well as other cloud storage via the mc client.

Reviewer: ```suggestion
Then, you can freely manage files and folders by copying, moving, adding and deleting between your local disk and JuiceFS storage as well as other cloud storage via the mc client.
```

    --from-literal=secret-key=<SECRET_KEY>
```

Here we have:

> - `name`: The JuiceFS file system name.

Reviewer: ```suggestion
- `name`: the name of the JuiceFS file system.
```

> - `metaurl`: URL for connecting metadata engine (e.g. Redis). Read [this document](../reference/how_to_setup_metadata_engine.md) for more information.

Reviewer: 直接说`the URL of the metadata database engine`会不会更好些？

> - `storage`: Object storage type, such as `s3`, `gs`, `oss`. Read [this document](../reference/how_to_setup_object_storage.md) for the full supported list.

Reviewer: ```suggestion
- `storage`: Object storage type, such as `s3`, `gs`, `oss`. Read [this document](../reference/how_to_setup_object_storage.md) to find all supported object storages.
```

- `bucket`: Bucket URL. Read [this document](../reference/how_to_setup_object_storage.md) to learn how to setup different object storage.
- `access-key`: Access key of object storage. Read [this document](../reference/how_to_setup_object_storage.md) for more information.
- `secret-key`: Secret key of object storage. Read [this document](../reference/how_to_setup_object_storage.md) for more information.

> Then download the S3 gateway [deployment YAML](https://github.com/juicedata/juicefs/blob/main/deploy/juicefs-s3-gateway.yaml) and create the `Deployment` and `Service` resources with `kubectl`. The following points require special attention:

Reviewer: ```suggestion
Then download the S3 gateway [deployment YAML](https://github.com/juicedata/juicefs/blob/main/deploy/juicefs-s3-gateway.yaml) and create the `Deployment` and `Service` resources with `kubectl`. Please pay special attention to:
```

- Please replace `${NAMESPACE}` in the following command with the Kubernetes namespace of the actual S3 gateway deployment, which defaults to `kube-system`.

> - The `replicas` for `Deployment` defaults to 1, please adjust as appropriate.

Reviewer: ```suggestion
- The `replicas` for `Deployment` defaults to 1, and please adjust as needed.
```

> - The latest version of `juicedata/juicefs-csi-driver` image is used by default, which has already integrated the latest version of JuiceFS client, please check [here](https://github.com/juicedata/juicefs-csi-driver/releases) for the specific integrated JuiceFS client version.

Reviewer: ```suggestion
- The latest version of `juicedata/juicefs-csi-driver` image is used by default, which has already integrated the latest version of JuiceFS client. Please check [here](https://github.com/juicedata/juicefs-csi-driver/releases) for the specific integrated JuiceFS client version.
```

- The `initContainers` of `Deployment` will first try to format the JuiceFS file system, if you have already formatted it in advance, this step will not affect the existing JuiceFS file system.

>   - The default port number that the S3 gateway listens on is 9000

Reviewer: ```suggestion
- The default port number that the S3 gateway listens on is 9000
```

> - The [startup options](../reference/command_reference.md#juicefs-gateway) of S3 gateway are the default values, please adjust them according to your actual needs.

Reviewer: ```suggestion
- The [startup options](../reference/command_reference.md#juicefs-gateway) of S3 gateway will use default values if not specified.
```

1. Prepare a YAML file

>    Create a configuration file, for example: `values.yaml`. Copy and fill the placeholders of the following configuration information. Among them, the `secret` part is the information related to the JuiceFS file system, you can refer to [JuiceFS Quick Start Guide](../getting-started/for_local.md) for more information.

Reviewer: ```suggestion
   Create a configuration file, for example: `values.yaml`. Copy and fill in the following configuration information. Among them, the `secret` part is the information related to the JuiceFS file system, and you can refer to [JuiceFS Quick Start Guide](../getting-started/for_local.md) for more information.
```

     bucket: "<bucket>"
   ```

>    If you want to deploy Ingress, add this snippet into `values.yaml`:

Reviewer: ```suggestion
   If you want to deploy Ingress, add the following snippet into `values.yaml`:
```

## Use a full-featured S3 gateway

If you need to use some advanced features of the MinIO S3 gateway, you can pull [the gateway branch of this repository](https://github.com/juicedata/minio/tree/gateway) and compile MinIO yourself. This branch is developed based on [MinIO RELEASE.2022-03-05T06-32-39Z](https://github.com/minio/minio/tree/RELEASE.2022-03-05T06-32-39Z) with JuiceFS gateway supported, which supports full functionality of MinIO gateways such as [multi-user management](https://docs.min.io/docs/minio-multi-user-quickstart-guide.html) while using JuiceFS as a backend.

### Compile

:::tip
This branch relies on a newer version of JuiceFS. Please refer to the [`go.mod`](https://github.com/juicedata/minio/blob/gateway/go.mod) file for the specific JuiceFS version.

> Similar to [manually compiling JuiceFS client](../getting-started/installation.md#manually-compiling), you need to install some dependencies in advance to compile S3 gateway normally.

Reviewer: ```suggestion
Similar to [manually compiling JuiceFS client](../getting-started/installation.md#manually-compiling), you need to install some dependencies in advance to compile S3 gateway.
```

$ ./minio gateway juicefs --console-address ':59001' redis://localhost:6379
```

> The port number of the S3 gateway console is explicitly specified here as 59001. If not specified, a port will be randomly assigned. According to the command line prompt, open the [http://127.0.0.1:59001](http://127.0.0.1:59001) address in the browser to access the console, as shown in the following figure:

Reviewer: ```suggestion
The port number of the S3 gateway console is explicitly specified here as 59001. If not specified, a port will be randomly assigned. According to the command line prompt, open the  address [http://127.0.0.1:59001](http://127.0.0.1:59001) in the browser to access the console, as shown in the following snapshot:
```
